extensions/sql-database-projects/src/common/utils.ts

```diff
@@ -143,3 +143,74 @@ export function readSqlCmdVariables(xmlDoc: any): Record<string, string> {
 
 	return sqlCmdVariables;
 }
+
+/**
+ * 	Removes $() around a sqlcmd variable
+ * @param name
+ */
+export function removeSqlCmdVariableFormatting(name: string | undefined): string {
+	if (!name || name === '') {
+		return '';
+	}
+
+	if (name.length > 3) {
+		// Trim in case we get "  $(x)"
+		name = name.trim();
+		let indexStart = name.startsWith('$(') ? 2 : 0;
+		let indexEnd = name.endsWith(')') ? 1 : 0;
+		if (indexStart > 0 || indexEnd > 0) {
+			name = name.substr(indexStart, name.length - indexEnd - indexStart);
+		}
+	}
+
+	// Trim in case the customer types "  $(x   )"
+	return name.trim();
+}
+
+/**
+ * 	Format as sqlcmd variable by adding $() if necessary
+ * if the variable already starts with $(, then add )
+ * @param name
+ */
+export function formatSqlCmdVariable(name: string): string {
+	// TODO: check if it's a valid variable name and just return if it isn't
+	if (!name || name === '') {
+		return name;
+	}
+
+	// Trim in case we get "  $(x)"
+	name = name.trim();
+
+	if (!name.startsWith('$(') && !name.endsWith(')')) {
+		name = `$(${name})`;
+	} else if (name.startsWith('$(') && !name.endsWith(')')) {
+		// add missing end parenthesis, same behavior as SSDT
+		name = `${name})`;
+	}
+
+	return name;
+}
+
+/**
+ * Checks if it's a valid sqlcmd variable name
+ * https://docs.microsoft.com/en-us/sql/ssms/scripting/sqlcmd-use-with-scripting-variables?redirectedfrom=MSDN&view=sql-server-ver15#guidelines-for-scripting-variable-names-and-values
+ * @param name variable name to validate
+ */
+export function isValidSqlCmdVariableName(name: string | undefined): boolean {
```

Comment:
If this throws instead of returning a boolean, you can give an explanation to the user as to why their variable name isn't valid.

Reply:
created https://github.com/microsoft/azuredatastudio/issues/12231. For now, this is just the same as what SSDT does for invalid variable names